docs/compiler_api.md

```diff
@@ -0,0 +1,121 @@
+# Compiler API
+
+## Background
+
+We assume readers are familiar with at least the basics of [sharding representation](sharding_representation.md), which describes how the sharding of a tensor can be expressed in Shardy. This document shows how the sharding representations can be used in a program, e.g. to attach a sharding to a specific tensor of the program.
+
+Sharding propagation is the process of deciding on a sharding for every tensor in a program given sharding constraints for a subset of the tensors. Shardy’s compiler API exposes several ways to influence/control sharding propagation. Additionally it allows users to insert manually sharded computations into their programs.
+
+## Objective
+
+This doc describes the design of such API components in Shardy and explains their behavior and invariants. Note that while this API is used to control sharding propagation, this doc is **NOT** going to discuss anything about the behavior of propagation nor how it’s designed.
+
+## Overview
+
+* [**Input/output shardings**](#inputoutput-shardings) - attach a sharding to an input or output of the main function, to indicate that this is how the input/output tensor should be sharded when given-to/returned-from the function.
+
+* [**Sharding Constraint**](#sharding-constraint) - attach a sharding to an intermediate tensor (e.g. the result of a matmul) to indicate that this is how that tensor, or a subset of its uses, should be sharded.
+
+* [**Shard As**](#shard-as) - group multiple tensors by an ID to indicate that they should be sharded in the same way.
```

Comment:
Since the op is called `ShardingGroupOp`, maybe lets refer to this as `Sharding Group`?